sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -286,44 +329,51 @@ class RocksDB(
    */
   def commit(): Long = {
     val newVersion = loadedVersion + 1
-    val checkpointDir = createTempDir("checkpoint")
-    var rocksDBBackgroundThreadPaused = false
     try {
-      // Make sure the directory does not exist. Native RocksDB fails if the directory to
-      // checkpoint exists.
-      Utils.deleteRecursively(checkpointDir)
 
       logInfo(s"Flushing updates for $newVersion")
-      val flushTimeMs = timeTakenMs { db.flush(flushOptions) }
-
-      val compactTimeMs = if (conf.compactOnCommit) {
-        logInfo("Compacting")
-        timeTakenMs { db.compactRange() }
-      } else 0
-
-      logInfo("Pausing background work")
-      val pauseTimeMs = timeTakenMs {
-        db.pauseBackgroundWork() // To avoid files being changed while committing
-        rocksDBBackgroundThreadPaused = true
-      }
 
-      logInfo(s"Creating checkpoint for $newVersion in $checkpointDir")
-      val checkpointTimeMs = timeTakenMs {
-        val cp = Checkpoint.create(db)
-        cp.createCheckpoint(checkpointDir.toString)
+      var flushTimeMs = 0L
+      var checkpointTimeMs = 0L
+      if (shouldCreateSnapshot()) {
+        // Need to flush the change to disk before creating a checkpoint
+        // because rocksdb wal is disabled.
+        flushTimeMs = timeTakenMs { db.flush(flushOptions) }
+        checkpointTimeMs = timeTakenMs {
+          val checkpointDir = createTempDir("checkpoint")
+          // Make sure the directory does not exist. Native RocksDB fails if the directory to
+          // checkpoint exists.
+          Utils.deleteRecursively(checkpointDir)
+          val cp = Checkpoint.create(db)
+          cp.createCheckpoint(checkpointDir.toString)
+          synchronized {
+            latestCheckpoint.foreach(_.close())
```

Comment:
Can we move these into separate functions where `latestCheckpoint` is being modified ?

Reply:
It is difficult to refactor this into a helper function because `latestCheckpoint` is accessed in slightly different pattern in different places.

Reply:
Could we not move this function into something like

```
private def updateLatestCheckpoint(newVersion: Long, checkpointDir: String): Unit = synchronized {
            latestCheckpoint.foreach(_.close())
            latestCheckpoint = Some(
              RocksDBCheckpoint(checkpointDir, newVersion, numKeysOnWritingVersion))
            lastCheckpointVersion = newVersion
}
```

Same for other places ?

Reply:
We can move this part out to a helper function, but it have only 3 lines of code and can not be reused anywhere, I don't think it improves the readability.

Reply:
I meant refactor such that all references to `latestCheckpoint` are modified within function(s) ?

Reply:
I thought about it from the beginning. But in the 3 places` latestCheckpoint` is modified, they are in slightly different pattern and can not fit inside a helper function.